File: demo/android/app/src/main/java/com/xprobe/thunderdemo/rpc/DemoInvocation.java
package com.xprobe.thunderdemo.rpc;

import android.content.Context;

import com.xprobe.rpc.ChannelCache;
import com.xprobe.rpc.CustomInvocation;
import com.xprobe.thunderdemo.rtc.RtcManager;

import java.util.HashMap;
import java.util.Map;

/**
 * RPC 自定义调用方案（ver=2）：
 * <p>
 * 测试端下发 {"api":{"apiName":"命令名","params":{...}},"threadMode":0,"ver":2}，
 * 本类将命令映射到 RtcManager 的简易操作并回传返回值，配合 Python 客户端
 * 即可一键完成「初始化 -> 进频道 -> 订阅 -> 播放」的自动化验证。
 * <p>
 * 支持命令：
 * <ul>
 *   <li>createEngine   {appId, sceneId}          初始化 SDK，返回耗时(ms)</li>
 *   <li>destroyEngine  {}                        销毁 SDK，返回 0</li>
 *   <li>joinRoom       {roomName, uid}           进频道，返回错误码</li>
 *   <li>leaveRoom      {}                        退频道，返回错误码</li>
 *   <li>addSubscribe   {roomName, uid}           订阅远端音视频，返回错误码</li>
 *   <li>removeSubscribe{roomName, uid}           取消订阅，返回错误码</li>
 *   <li>startLocalPreview {}                     开启本地采集与预览，返回错误码</li>
 *   <li>stopLocalPreview  {}                     关闭本地采集与预览，返回错误码</li>
 *   <li>setupRemoteVideo  {uid}                  设置远端视频画布，返回错误码</li>
 *   <li>getState      {}                         返回 {init, room, uid, remoteUid}</li>
 *   <li>echo          {任意}                     回显 apiName 与 params（协议回归）</li>
 *   <li>scheduleCallback {name, info, delayMs}   先 return 再异步 callback（协议回归）</li>
 * </ul>
 */
public class DemoInvocation implements CustomInvocation {

    private final Context mContext;
    private final android.os.Handler mMainHandler =
            new android.os.Handler(android.os.Looper.getMainLooper());

    public DemoInvocation(Context context) {
        mContext = context.getApplicationContext();
    }

    @Override
    public void callMethod(String api, Map<String, Object> params) {
        if (params == null) {
            params = new HashMap<>();
        }
        try {
            dispatch(api, params);
        } catch (Exception e) {
            ChannelCache.sendErrorInfo(api, android.util.Log.getStackTraceString(e));
        }
    }

    private void dispatch(String api, Map<String, Object> params) {
        RtcManager rtc = RtcManager.getInstance();
        if ("createEngine".equals(api)) {
            String appId = str(params.get("appId"), "10034");
            long sceneId = lng(params.get("sceneId"), 0);
            long cost = rtc.initialize(mContext, appId, sceneId);
            ChannelCache.sendMethodReturn(api, String.valueOf(cost));
        } else if ("destroyEngine".equals(api)) {
            rtc.deInitialize();
            ChannelCache.sendMethodReturn(api, "0");
        } else if ("joinRoom".equals(api)) {
            String room = str(params.get("roomName"), "82552971");
            String uid = str(params.get("uid"), "123456789");
            int ret = rtc.joinRoom(room, uid);
            ChannelCache.sendMethodReturn(api, String.valueOf(ret));
        } else if ("leaveRoom".equals(api)) {
            int ret = rtc.leaveRoom();
            ChannelCache.sendMethodReturn(api, String.valueOf(ret));
        } else if ("addSubscribe".equals(api)) {
            String room = str(params.get("roomName"), rtc.getRoomName());
            String uid = str(params.get("uid"), rtc.getRemoteUid());
            int ret = rtc.addSubscribe(room, uid);
            ChannelCache.sendMethodReturn(api, String.valueOf(ret));
        } else if ("removeSubscribe".equals(api)) {
            String room = str(params.get("roomName"), rtc.getRoomName());
            String uid = str(params.get("uid"), rtc.getRemoteUid());
            int ret = rtc.removeSubscribe(room, uid);
            ChannelCache.sendMethodReturn(api, String.valueOf(ret));
        } else if ("startLocalPreview".equals(api)) {
            int ret = rtc.startLocalPreview();
            ChannelCache.sendMethodReturn(api, String.valueOf(ret));
        } else if ("stopLocalPreview".equals(api)) {
            int ret = rtc.stopLocalPreview();
            ChannelCache.sendMethodReturn(api, String.valueOf(ret));
        } else if ("setupRemoteVideo".equals(api)) {
            String uid = str(params.get("uid"), rtc.getRemoteUid());
            int ret = rtc.setupRemoteVideo(uid);
            ChannelCache.sendMethodReturn(api, String.valueOf(ret));
        } else if ("getState".equals(api)) {
            String state = "init=" + rtc.isInitialized()
                    + ", room=" + rtc.getRoomName()
                    + ", uid=" + rtc.getUid()
                    + ", remoteUid=" + rtc.getRemoteUid();
            ChannelCache.sendMethodReturn(api, state);
        } else if ("echo".equals(api)) {
            ChannelCache.sendMethodReturn(api, "echo api=" + api + " params=" + params);
        } else if ("scheduleCallback".equals(api)) {
            final String name = str(params.get("name"), "onProbeCallback");
            final String info = str(params.get("info"), "ok");
            long delayMs = lng(params.get("delayMs"), 200);
            ChannelCache.sendMethodReturn(api, "0");
            mMainHandler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    ChannelCache.sendCallbackMsg(name, info);
                }
            }, delayMs);
        } else {
            ChannelCache.sendErrorInfo(api, "unknown api: " + api);
        }
    }

    private static String str(Object o, String def) {
        return o == null ? def : String.valueOf(o);
    }

    private static long lng(Object o, long def) {
        if (o == null) {
            return def;
        }
        try {
            if (o instanceof Number) {
                return ((Number) o).longValue();
            }
            return Long.parseLong(String.valueOf(o));
        } catch (NumberFormatException e) {
            return def;
        }
    }
}

File: demo/pc/rpc/DemoInvocation.cpp
#include "rpc/DemoInvocation.h"

#include <chrono>
#include <cstdint>
#include <thread>

#include "rtc/RtcManager.h"

namespace {

std::string strParam(const xprobe::JsonValue& params, const char* key, const std::string& def) {
    const xprobe::JsonValue* v = params.find(key);
    if (v == nullptr || v->isNull()) {
        return def;
    }
    if (v->isString()) {
        return v->asString();
    }
    return v->serialize();
}

int64_t lngParam(const xprobe::JsonValue& params, const char* key, int64_t def) {
    const xprobe::JsonValue* v = params.find(key);
    if (v == nullptr || v->isNull()) {
        return def;
    }
    if (v->isNumber()) {
        return v->asInt(def);
    }
    return def;
}

} // namespace

DemoInvocation::DemoInvocation(xprobe::AutoTestMgr* mgr) : mgr_(mgr) {}

void DemoInvocation::callMethod(const std::string& apiName, const xprobe::JsonValue& params) {
    try {
        if (!apiName.empty() && apiName.front() == '{') {
            handleLegacyPassthrough(apiName);
            return;
        }
        dispatch(apiName, params);
    } catch (const std::exception& e) {
        mgr_->sendError(apiName, e.what());
    } catch (...) {
        mgr_->sendError(apiName, "unknown exception");
    }
}

void DemoInvocation::handleLegacyPassthrough(const std::string& apiName) {
    // C++ Core 无反射：ver!=2 / field 命令仍透传到 CustomInvocation，供协议回归。
    mgr_->sendReturn("legacyApi", apiName);
}

void DemoInvocation::dispatch(const std::string& api, const xprobe::JsonValue& params) {
    RtcManager& rtc = RtcManager::getInstance();

    if (api == "createEngine") {
        std::string appId = strParam(params, "appId", "10034");
        long sceneId = static_cast<long>(lngParam(params, "sceneId", 0));
        long cost = rtc.initialize(appId, sceneId);
        mgr_->sendReturn(api, std::to_string(cost));
    } else if (api == "destroyEngine") {
        rtc.deInitialize();
        mgr_->sendReturn(api, "0");
    } else if (api == "joinRoom") {
        std::string room = strParam(params, "roomName", "82552971");
        std::string uid = strParam(params, "uid", "123456789");
        std::string token = strParam(params, "token", "");
        int ret = rtc.joinRoom(room, uid, token);
        mgr_->sendReturn(api, std::to_string(ret));
    } else if (api == "leaveRoom") {
        int ret = rtc.leaveRoom();
        mgr_->sendReturn(api, std::to_string(ret));
    } else if (api == "addSubscribe") {
        std::string room = strParam(params, "roomName", rtc.getRoomName());
        std::string uid = strParam(params, "uid", rtc.getRemoteUid());
        int ret = rtc.addSubscribe(room, uid);
        mgr_->sendReturn(api, std::to_string(ret));
    } else if (api == "removeSubscribe") {
        std::string room = strParam(params, "roomName", rtc.getRoomName());
        std::string uid = strParam(params, "uid", rtc.getRemoteUid());
        int ret = rtc.removeSubscribe(room, uid);
        mgr_->sendReturn(api, std::to_string(ret));
    } else if (api == "startLocalPreview") {
        int ret = rtc.startLocalPreview();
        mgr_->sendReturn(api, std::to_string(ret));
    } else if (api == "stopLocalPreview") {
        int ret = rtc.stopLocalPreview();
        mgr_->sendReturn(api, std::to_string(ret));
    } else if (api == "setupRemoteVideo") {
        std::string uid = strParam(params, "uid", rtc.getRemoteUid());
        int ret = rtc.setupRemoteVideo(uid);
        mgr_->sendReturn(api, std::to_string(ret));
    } else if (api == "getState") {
        std::string state = "init=" + std::string(rtc.isInitialized() ? "true" : "false")
            + ", room=" + rtc.getRoomName()
            + ", uid=" + rtc.getUid()
            + ", remoteUid=" + rtc.getRemoteUid();
        mgr_->sendReturn(api, state);
    } else if (api == "echo") {
        mgr_->sendReturn(api, "echo api=" + api + " params=" + params.serialize());
    } else if (api == "scheduleCallback") {
        std::string name = strParam(params, "name", "onProbeCallback");
        std::string info = strParam(params, "info", "ok");
        long delayMs = lngParam(params, "delayMs", 200);
        mgr_->sendReturn(api, "0");
        xprobe::AutoTestMgr* mgr = mgr_;
        std::thread([mgr, name, info, delayMs]() {
            if (delayMs > 0) {
                std::this_thread::sleep_for(std::chrono::milliseconds(delayMs));
            }
            mgr->sendCallback(name, info);
        }).detach();
    } else {
        mgr_->sendError(api, "unknown api: " + api);
    }
}
